cores/esp32/esp32-hal-cpu.c

```diff
@@ -37,54 +38,110 @@ const uint32_t MHZ = 1000000;
 static apb_change_t * apb_change_callbacks = NULL;
 static xSemaphoreHandle apb_change_lock = NULL;
 
+void IRAM_ATTR twiddle( const char* pattern){
+int len=strlen(pattern),pos=0;
+digitalWrite(D_A,HIGH);
+digitalWrite(D_B,LOW);
+while(pos<len){
+  if(pattern[pos++]=='1'){
+    digitalWrite(D_B,HIGH);
+  } else {
+    digitalWrite(D_B,LOW);
+  }
+  digitalWrite(D_B,LOW);
+}
+digitalWrite(D_A,LOW);
+digitalWrite(D_B,LOW);
+}  
+
+
 static void initApbChangeCallback(){
     static volatile bool initialized = false;
     if(!initialized){
-        initialized = true;
         apb_change_lock = xSemaphoreCreateMutex();
-        if(!apb_change_lock){
-            initialized = false;
-        }
+        initialized = apb_change_lock != NULL;
     }
 }
 
-static void triggerApbChangeCallback(apb_change_ev_t ev_type, uint32_t old_apb, uint32_t new_apb){
+static bool triggerApbChangeCallback(apb_change_ev_t ev_type, uint32_t old_apb, uint32_t new_apb){
     initApbChangeCallback();
     xSemaphoreTake(apb_change_lock, portMAX_DELAY);
     apb_change_t * r = apb_change_callbacks;
-    while(r != NULL){
-        r->cb(r->arg, ev_type, old_apb, new_apb);
-        r=r->next;
+    bool success = true;
+    while((r != NULL)&& success){
+/*
+        switch(ev_type){
+ 
+            case APB_BEFORE_CHANGE:
+                twiddle("1100");
+                break;
+
+            case APB_AFTER_CHANGE:
+                twiddle("1000");
+                break;
+            default :;
+        }
+*/
+//        log_i("callback %p(%d)",r->cb,ev_type);
+        success = r->cb(r->arg, ev_type, old_apb, new_apb);
+        if(ev_type == APB_AFTER_CHANGE){
+            r = r->next;
+            success = true; // don't care after processor clock change
+        } else { //APB_BEFORE_CHANGE
+            if(success){
+                r = r->next;
+            }
+        }
+    }
+    if(!success && ev_type == APB_BEFORE_CHANGE){
+        if(r != NULL) r=r->prev;
+        while(r != NULL){
+            r->cb(r->arg, APB_ABORT_CHANGE, old_apb, new_apb);
+            r = r->prev;
+        }
     }
+/*        
+        switch(ev_type){
+            case APB_BEFORE_CHANGE:
+                twiddle("1101");
+                break;
+
+            case APB_AFTER_CHANGE:
+                twiddle("1001");
+                break;
+            default :;
+        }
+*/
     xSemaphoreGive(apb_change_lock);
+    return success;
+
 }
 
 bool addApbChangeCallback(void * arg, apb_change_cb_t cb){
     initApbChangeCallback();
-    apb_change_t * c = (apb_change_t*)malloc(sizeof(apb_change_t));
```

Comment:
why is this whole change necessary? to speed up? how many calls will be made to this function?

Reply:
I just hate `goto` and this flow is easier to understand, 
* lock
* check duplicate
 add if new
* unlock

the double linked list may not be necessary,  I just added it while testing to see if execution calling order made a difference,  I was running into deadlocks because of `log_x()` calls during debuging